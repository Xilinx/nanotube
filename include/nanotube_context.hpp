/**************************************************************************\
*//*! \file nanotube_context.hpp
** \author  Neil Turton <neilt@amd.com>
**  \brief  Implementation of Nanotube contexts.
**   \date  2020-09-01
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/

#ifndef NANOTUBE_CONTEXT_HPP
#define NANOTUBE_CONTEXT_HPP

#include "nanotube_api.h"

#include <map>
#include <unordered_map>
#include <utility>

class nanotube_thread;

/*! A Nanotube context for holding resources.
 *
 * A Nanotube context is used for passing resources to a packet kernel
 * or a thread.
 *
 * This class needs to be merged with nanotube_context, but that
 * struct is currently visible to the kernel written in C.
 */
class nanotube_context
{
public:
  nanotube_context();
  ~nanotube_context();

  /*! Bind a thread to a context. */
  void bind_thread(nanotube_thread *thread);

  /*! Unbind a thread from a context. */
  void unbind_thread(nanotube_thread *thread);

  /*! Determine whether the associated thread is the current thread. */
  bool is_current();

  /*! Determine whether the associated thread is stopped. */
  bool is_stopped();

  /*! Add a channel to the context.
   *
   * \param channel_id The ID which will be used to access the channel.
   * \param channel    The channel to add.
   * \param flags      The types of access which need to be supported.
   */
  void add_channel(nanotube_channel_id_t channel_id,
                   nanotube_channel* channel,
                   nanotube_channel_flags_t flags);

  /*! Retrieve a channel from the context.
   *
   * \param channel_id  The ID which was passed to add_channel.
   * \param flags       The type of access which is requested.
   *
   * \return            A reference to the channel.
   *
   * This method raised an error on failure.
   */
  nanotube_channel &find_channel(nanotube_channel_id_t channel_id,
                                 nanotube_channel_flags_t flags);

  /*!
   * Add a map to the context.
   *
   * \param id    Numerical ID of the map to be added.
   * \param map   Pointer to the map that should be added.
   */
  void add_map(nanotube_map_id_t id, nanotube_map_t* map);

  /*!
   * Get the specified map; returns nullptr in case the id was not found.
   * \param id    The ID of the map that should be returned.
   * \return      Pointer to the map.
   */
  nanotube_map_t* get_map(nanotube_map_id_t id);

  /*! Check that this context can be used from the current thread. 
   *
   * This method raises a fatal error if the context cannot be used
   * from the current thread.
   */
  void check_thread();

  /*! Wake up the thread associated with this context. */
  void wake();

  nanotube_thread* get_thread() {return m_thread;}

private:
  typedef std::pair<nanotube_channel_id_t, nanotube_channel_flags_t>
    port_info_t;

  // The thread associated with the context, if any.
  nanotube_thread *m_thread;

  std::map<port_info_t, nanotube_channel_t *>            m_channels;
  std::unordered_map<nanotube_map_id_t, nanotube_map_t*> m_maps;
};

#endif // NANOTUBE_CONTEXT_HPP
