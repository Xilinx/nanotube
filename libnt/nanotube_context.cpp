/**************************************************************************\
*//*! \file nanotube_context.cpp
** \author  Neil Turton <neilt@amd.com>
**  \brief  Implementation of Nanotube contexts.
**   \date  2020-09-01
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/

#include "nanotube_context.hpp"

#include "nanotube_api.h"
#include "nanotube_channel.hpp"
#include "nanotube_thread.hpp"
#include "processing_system.hpp"

#include <cassert>

///////////////////////////////////////////////////////////////////////////

nanotube_context::nanotube_context():
  m_thread(nullptr)
{
}

nanotube_context::~nanotube_context()
{
  assert(m_thread == nullptr);
}

void nanotube_context::bind_thread(nanotube_thread *thread)
{
  assert(m_thread == nullptr);
  m_thread = thread;
}

void nanotube_context::unbind_thread(nanotube_thread *thread)
{
  assert(m_thread == thread);
  m_thread = nullptr;
}

bool nanotube_context::is_current()
{
  if (m_thread == nullptr)
    return false;
  return m_thread->is_current();
}

bool nanotube_context::is_stopped()
{
  if (m_thread == nullptr)
    return true;
  return m_thread->is_stopped();
}

void nanotube_context::add_channel(nanotube_channel_id_t channel_id,
                                        nanotube_channel_t* channel,
                                        nanotube_channel_flags_t flags)
{
  if (flags & NANOTUBE_CHANNEL_READ) {
    auto res = m_channels.emplace(
      std::make_pair(channel_id, NANOTUBE_CHANNEL_READ), channel);
    if (!res.second)
      throw std::runtime_error("Channel is already registered for"
                               " reading.");
    channel->set_reader(this);
  }

  if (flags & NANOTUBE_CHANNEL_WRITE) {
    auto res = m_channels.emplace(
      std::make_pair(channel_id, NANOTUBE_CHANNEL_WRITE), channel);
    if (!res.second)
      throw std::runtime_error("Channel is already registered for"
                               " writing.");
    channel->set_writer(this);
  }
}

nanotube_channel_t &
nanotube_context::find_channel(nanotube_channel_id_t channel_id,
                                    nanotube_channel_flags_t flags)
{
  auto it = m_channels.find(std::make_pair(channel_id, flags));
  assert (it != m_channels.end());
  return *(it->second);
}

void
nanotube_context::add_map(nanotube_map_id_t id, nanotube_map_t* map) {
  m_maps[id] = map;
}

nanotube_map_t*
nanotube_context::get_map(nanotube_map_id_t id) {
  auto it = m_maps.find(id);
  if( it == m_maps.end() )
    return nullptr;
  return it->second;
}

void nanotube_context::check_thread()
{
  assert(m_thread != nullptr);
  m_thread->check_current();
}

void nanotube_context::wake()
{
  assert(m_thread != nullptr);
  m_thread->wake();
}

///////////////////////////////////////////////////////////////////////////

nanotube_context_t *nanotube_context_create()
{
  auto* ctx = new nanotube_context();
  std::unique_ptr<nanotube_context> context_ptr(ctx);

  processing_system &ps = processing_system::get_current();
  ps.add_context(std::move(context_ptr));
  return ctx;
}

void nanotube_context_add_channel(nanotube_context_t* context,
                                  nanotube_channel_id_t channel_id,
                                  nanotube_channel_t* channel,
                                  nanotube_channel_flags_t flags)
{
  context->add_channel(channel_id, channel, flags);
}

void nanotube_context_add_map(nanotube_context_t* context,
                              nanotube_map_t*     map) {
  context->add_map(map->id, map);
}

///////////////////////////////////////////////////////////////////////////
