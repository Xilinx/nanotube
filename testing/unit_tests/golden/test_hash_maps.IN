###########################################################################
# Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
# SPDX-License-Identifier: MIT
###########################################################################
nanotube_map: 123 0 8 64
key: de ad be ef fe ed ba cc
value: 00 00 11 11 22 22 33 33 44 44 55 55 66 66 77 77
       88 88 99 99 aa aa bb bb cc cc dd dd ee ee ff ff
       cc 01 cc 02 cc 03 cc 04 cc 05 cc 06 cc 07 cc 08
       09 cc 0a cc 0b cc 0c cc 0d cc 0e cc 0f cc 10 cc
end

nanotube_map: 999 0 4 32
key: de ad be ef
value: f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0
       f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0
key: fe ed ba cc
value: cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb
       cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb
key: c0 ff ee 00
value: 66 66 66 66 66 66 66 66 66 66 66 66 66 66 66 66
       66 66 66 66 66 66 66 66 66 66 66 66 66 66 66 66
end
