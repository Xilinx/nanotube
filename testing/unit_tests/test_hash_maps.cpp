/*******************************************************/
/*! \file test_hash_maps.cpp
** \author Stephan Diestelhorst <stephand@amd.com>
**  \brief Unit tests for Nanotube hash map implementation.
**   \date 2020-03-04
*//******************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "nanotube_api.h"
#include "nanotube_context.hpp"
#include "test.hpp"

static void print_bytes(uint8_t* data, size_t len) {
  for( size_t i = 0; i < len; ++i ) {
    printf("%02x ", data[i]);
    if( (i % 8) == 7 )
      printf("\n");
  }
}

/*!
** Sentinel / canary entries and separate map.
**
** The sentinel map and sentinel map entries serve as canaries to detect
** modifications of other maps / map entries than those explicitly manipulated
** in the test.  These modifications could occur due to bad pointer arithmetic
** in the map implementation, stray writes, and use-after-free.
**
** We create a set of sentinel keys + value and put those into the "work" map
** (under keys that are different than the "work" key from the test), and also
** use the same keys and values (in a different mapping) in a separate
** sentinel map.
**/
#define SENTINELS (10)            /*< Number of sentinel entries to use */
#define SENTINEL_MAP (0xfee7)     /*< ID of the sentinel map */
static nanotube_map_t* sentinel_map;

/* Golden keys and data for canary / sentinel checking */
static uint8_t sentinel_key[SENTINELS][8];
static uint8_t sentinel_data[SENTINELS][64];


/*!
** Generate sentinel / canary data.
**
** Generate sentinel constant data (keys & value) and insert into the work map
** specified by map, and also into a separate map.  Neither keys in the work
** map, nor the content of the separate map should be modified by the test
** application.
**
** \param map ID of the work map that is being tested
**/
static void add_sentinel_data(nanotube_context_t* ctx,
                              nanotube_map_id_t work_map) {
  int len;
  sentinel_map = nanotube_map_create(SENTINEL_MAP, NANOTUBE_MAP_TYPE_HASH,
                                     8, 64);
  nanotube_context_add_map(ctx, sentinel_map);

  for( int i = 0; i < SENTINELS; ++i ) {
    /* Generate "random" keys */
    for ( size_t j = 0; j < sizeof(sentinel_key[0]); ++j )
      sentinel_key[i][j] = 0x12 + i + j;
    /* Generate "random" values */
    for ( size_t j = 0; j < sizeof(sentinel_data[0]); ++j )
      sentinel_data[i][j] = i ^ j;
  }

  /* Add key->value to the work map and to the separate sentinel map */
  for ( int i = 0; i < SENTINELS; ++i ) {
    /* In work map: key[i] -> value[i] */
    len = nanotube_map_write(ctx, work_map,
              sentinel_key[i], sizeof(sentinel_key[0]),
              sentinel_data[i], 0, sizeof(sentinel_data[0]));
    assert_eq(len, sizeof(sentinel_data[0]));
    /* Reverse mapping for sentinel map */
    unsigned i_rev = SENTINELS - 1 - i;
    len = nanotube_map_write(ctx, SENTINEL_MAP,
              sentinel_key[i_rev], sizeof(sentinel_key[0]),
              sentinel_data[i], 0, sizeof(sentinel_data[0]));
    assert_eq(len, sizeof(sentinel_data[0]));
  }
}

/*!
** Check that all sentinel data is still intact.
**
** \param map ID of work map that contains sentinel entries.
**/
static void check_sentinel_data(nanotube_context_t* ctx,
                                nanotube_map_id_t work_map) {
  int len;
  uint8_t data[64];

  for( int i = 0; i < SENTINELS; ++i ) {
    /* Read and check data */
    len = nanotube_map_read(ctx, work_map,
              sentinel_key[i], sizeof(sentinel_key[0]),
              data, 0, sizeof(data));
    assert_eq(len, sizeof(data));
    assert_eq(memcmp(data, sentinel_data[i], sizeof(data)), 0);
    /* Reverse mapping for sentinel map */
    len = nanotube_map_read(ctx, SENTINEL_MAP,
              sentinel_key[i], sizeof(sentinel_key[0]),
              data, 0, sizeof(data));
    assert_eq(len, sizeof(data));
    unsigned i_rev = SENTINELS - 1 - i;
    assert_eq(memcmp(data, sentinel_data[i_rev], sizeof(data)), 0);
  }
}

/*!
** Remove sentinel data from supplied map.
**/
void remove_sentinel_data(nanotube_context_t* ctx, nanotube_map_id_t map) {
  size_t len;
  for( int i = 0; i < SENTINELS; ++i) {
    len = nanotube_map_remove(ctx, map, sentinel_key[i],
                              sizeof(sentinel_key[0]));
    assert_eq(len==0, 0);
  }
}

/*!
** Test code for testing maps.
**
** \return Number of errors found.
**/
static void test_maps(void) {
  printf("\n\n_Testing_Maps_\n");
  auto* ctx = new nanotube_context();

  const nanotube_map_id_t A_id = 42;
  nanotube_map_t* map_A;
  map_A = nanotube_map_create(A_id, NANOTUBE_MAP_TYPE_HASH, 8, 64);
  nanotube_context_add_map(ctx, map_A);
  add_sentinel_data(ctx, A_id);
  check_sentinel_data(ctx, A_id);

  uint8_t all_enabled[8];
  memset(all_enabled, 0xff, 8);

  uint8_t key[8]  = {0xde, 0xad, 0xbe, 0xef, 0xfe, 0xed, 0xba, 0xcc};
  uint8_t data[64];
  for( size_t i = 0; i < sizeof(data); ++i )
    data[i] = i;
  size_t len;

  /* Map read / write testing */
  nanotube_map_t* maps[3];

  printf("Map input\n");
  for( int i = 0; i < 3; ++i ) {
    maps[i] = nanotube_map_create_from_stdin();
    printf("  Read map %hi from stdin\n", nanotube_map_get_id(maps[i]));
  }
  assert_eq(nanotube_map_get_id(maps[0]), 123);
  assert_eq(nanotube_map_get_id(maps[1]), 999);
  assert_eq(nanotube_map_get_id(maps[2]), NANOTUBE_MAP_ILLEGAL);

  printf("Map output\n");
  for( int i = 0; i < 3; ++i ) {
    if( maps[i] != NULL )
      nanotube_map_print(maps[i]);
  }

  /* Simple map write */
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key), data,
                                  all_enabled, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes:\n", len);
  print_bytes(data, 64); printf("\n");
  assert_eq(len, 64);

  /* Testing map read of existing value */
  uint8_t data2[64];
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), data2, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes:\n", len);
  print_bytes(data2, 64); printf("\n");
  assert_eq(memcmp(data, data2, 64), 0);

  /* Writing & reading from an offset */
  size_t pos = 16;
  data[pos]  = 0xab;
  len = nanotube_map_write(ctx, A_id, key, sizeof(key), data + pos, pos, 1);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes at pos %zu value %02x.\n", len, pos, data[pos]);
  assert_eq(len, 1);

  pos = 10;
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), data2 + pos, pos, 10);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes with offset %zu:\n", len, pos);
  print_bytes(data2, 64); printf("\n");
  assert_eq(len, 10);
  assert_eq(data2[16], 0xab);

  /* Writing with enables & offset */
  uint8_t data_splat[32];
  memset(data_splat, 0xff, 32);
  /* Alternating, growing on/off pattern */
  uint8_t fancy_enable[4] = {/*00110010b*/  50, /*00111100b*/ 60,
                             /*11000000b*/ 192, /*00111111b*/ 63};
  pos = 32;
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key), data_splat,
                                  fancy_enable, pos, sizeof(data_splat));
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes at pos %zu with fancy enable mask.\n", len, pos);
  assert_eq(len, sizeof(data_splat));

  pos = 0;
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), data2 + pos, pos,
                          sizeof(data2));
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes with offset %zu:\n", len, pos);
  print_bytes(data2, 64); printf("\n");
  assert_eq(len, sizeof(data2));

  /* Illegal reads */
  /* Wrong map */
  nanotube_map_id_t illegal_id = 66;
  assert((illegal_id != A_id) && (illegal_id != SENTINEL_MAP));
  len = nanotube_map_read(ctx, illegal_id, key, sizeof(key), data2, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes from map %i\n", len, illegal_id);
  assert_eq(len, 0);

  /* No key */
  len = nanotube_map_read(ctx, A_id, NULL, sizeof(key), data2, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes from NULL key\n", len);
  assert_eq(len, 0);

  /* No data */
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), NULL, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes to NULL data\n", len);
  assert_eq(len, 0);

  /* Missing key */
  uint8_t missing_key[8] = {1, 2, 3, 4, 5, 6, 7, 8};
  assert_eq(memcmp(missing_key, key, sizeof(key)) != 0, 1);
  len = nanotube_map_read(ctx, A_id, missing_key, sizeof(missing_key), data2,
                          0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes from missing key\n", len);
  assert_eq(len, 0);

  /* Too big key */
  len = nanotube_map_read(ctx, A_id, key, sizeof(key) + 3, data2, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes from too big key\n", len);
  assert_eq(len, 0);

  /* Too small key */
  len = nanotube_map_read(ctx, A_id, key, sizeof(key) - 3, data2, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes from too small key\n", len);
  assert_eq(len, 0);

  /* Too much data */
  uint8_t huge_data[80] = {0};;
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), huge_data, 0,
              sizeof(huge_data));
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes with request of %zu bytes\n", len,
         sizeof(huge_data));
  print_bytes(huge_data, sizeof(huge_data)); printf("\n");
  /* We expect to read one value if the return space is too big */
  assert_eq(len, 64);

  /* Offset too far */
  pos = 32;
  memset(data2, 0, sizeof(data2));
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), data2, pos,
              sizeof(data2));
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes from offset %zu and requesting %zu bytes\n", len,
         pos, sizeof(data2));
  print_bytes(data2, sizeof(data2)); printf("\n");
  /* We (for now) expect to read as much as we can. */
  assert_eq(len, 32);
  /* Check that the overwritten 0xff's are still there */
  assert_eq(data2[1], 0xff);
  assert_eq(data2[4], 0xff);
  assert_eq(data2[5], 0xff);

  /* Illegal writes */
  /* Wrong map */
  len = nanotube_map_write_masked(ctx, illegal_id, key, sizeof(key), data,
                                  all_enabled, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes to map %i\n", len, illegal_id);
  assert_eq(len, 0);

  /* No key */
  len = nanotube_map_write_masked(ctx, A_id, NULL, sizeof(key), data,
                                  all_enabled, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes to NULL key\n", len);
  assert_eq(len, 0);

  /* No data */
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key), NULL,
                                  all_enabled, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes of NULL data\n", len);
  assert_eq(len, 0);

  /* No enables */
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key), data, NULL, 0,
                                  64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes with NULL enables\n", len);
  assert_eq(len, 0);

  /* Too big key */
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key) + 3, data,
                                  all_enabled, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes to large key\n", len);
  assert_eq(len, 0);

  /* Too small key */
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key) - 3, data,
                                  all_enabled, 0, 64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes to small key\n", len);
  assert_eq(len, 0);

  /* Too much data (due to offset) */
  pos = 5;
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key), data,
                                  all_enabled, pos, 64);
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes to offset %zu with req size %u \n", len, pos, 64);
  assert_eq(len, 0);

  /* Too much data (due to length) */
  pos = 0;
  len = nanotube_map_write_masked(ctx, A_id, key, sizeof(key), huge_data,
                                  all_enabled, pos, sizeof(huge_data));
  check_sentinel_data(ctx, A_id);
  printf("Wrote %zu bytes with length %zu\n", len, sizeof(huge_data));
  assert_eq(len, 0);

  /* Test the different "special" writes: UPDATE and INSERT */
  /* UPDATE with a present key */
  pos = 0;
  memset(data, 0xf0, sizeof(data) / 2);
  len = nanotube_map_update_masked(ctx, A_id, key, sizeof(key), data + pos,
                                   all_enabled, pos, sizeof(data) / 2);
  check_sentinel_data(ctx, A_id);
  printf("Updated %zu bytes at pos %zu.\n", len, pos);
  assert_eq(len, sizeof(data) / 2);
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), data2 + pos, pos,
                          sizeof(data2));
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes with offset %zu:\n", len, pos);
  print_bytes(data2, 64); printf("\n");
  assert_eq(len, sizeof(data2));
  assert_eq(data2[0], 0xf0);

  /* UPDATE with a non-existing key */
  ++key[5];
  pos = 0;
  len = nanotube_map_update_masked(ctx, A_id, key, sizeof(key), data + pos,
                                   all_enabled, pos, sizeof(data) / 2);
  check_sentinel_data(ctx, A_id);
  printf("Updated %zu bytes at pos %zu with missing key.\n", len, pos);
  assert_eq(len, 0);

  len = nanotube_map_update_masked(ctx, A_id, key, sizeof(key), data + pos,
                                   all_enabled, pos, sizeof(data) / 2);
  check_sentinel_data(ctx, A_id);
  printf("Re-Updated %zu bytes at pos %zu with missing key.\n", len, pos);
  assert_eq(len, 0);
  --key[5];

  /* INSERT with a non-existing key */
  ++key[5];
  pos = 0;
  len = nanotube_map_insert_masked(ctx, A_id, key, sizeof(key), data + pos,
                                   all_enabled, pos, sizeof(data) / 2);
  check_sentinel_data(ctx, A_id);
  printf("Inserted %zu bytes at pos %zu with new key.\n", len, pos);
  assert_eq(len, sizeof(data) / 2);
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), data2 + pos, pos,
                          sizeof(data2));
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes with offset %zu:\n", len, pos);
  print_bytes(data2, 64); printf("\n");
  assert_eq(len, sizeof(data2));
  assert_eq(data2[0], 0xf0);

  /* INSERT with the same, now-existing key */
  len = nanotube_map_insert_masked(ctx, A_id, key, sizeof(key), data + pos,
                                   all_enabled, pos, sizeof(data) / 2);
  check_sentinel_data(ctx, A_id);
  printf("Re-inserted %zu bytes at pos %zu with (not so) new key.\n", len, pos);
  assert_eq(len, 0);

  /* Remove entry from the map */
  len = nanotube_map_remove(ctx, A_id, key, sizeof(key));
  printf("Removed newly inserted key, result %zu.\n", len);
  check_sentinel_data(ctx, A_id);
  assert_eq(len!=0, 1);

  /* And again .. remove entry from the map */
  len = nanotube_map_remove(ctx, A_id, key, sizeof(key));
  printf("Re-removed newly inserted key, result %zu.\n", len);
  check_sentinel_data(ctx, A_id);
  assert_eq(len, 0);

  /* Read from the removed key */
  len = nanotube_map_read(ctx, A_id, key, sizeof(key), data2 + pos, pos,
                          sizeof(data2));
  check_sentinel_data(ctx, A_id);
  printf("Read %zu bytes with offset %zu from removed key\n", len, pos);
  print_bytes(data2, 64); printf("\n");
  assert_eq(len, 0);

  /* And print out the final map */
  remove_sentinel_data(ctx, A_id);
  printf("Final map state:\n");
  nanotube_map_print(map_A);

  nanotube_map_destroy(map_A);
  nanotube_map_destroy(sentinel_map);
  delete ctx;
}

int main(int argc, char *argv[]) {
  test_init(argc, argv);
  test_maps();
  return test_fini();
}
/* vim: set ts=8 et sw=2 sts=2 tw=75: */
