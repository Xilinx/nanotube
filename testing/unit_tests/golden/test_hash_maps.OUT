

_Testing_Maps_
Map input
  Read map 123 from stdin
  Read map 999 from stdin
  Read map -1 from stdin
Map output
nanotube_map: 123 0 8 64
key: de ad be ef fe ed ba cc
value: 00 00 11 11 22 22 33 33 44 44 55 55 66 66 77 77
       88 88 99 99 aa aa bb bb cc cc dd dd ee ee ff ff
       cc 01 cc 02 cc 03 cc 04 cc 05 cc 06 cc 07 cc 08
       09 cc 0a cc 0b cc 0c cc 0d cc 0e cc 0f cc 10 cc
end

nanotube_map: 999 0 4 32
key: fe ed ba cc
value: cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb
       cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb cb
key: c0 ff ee 00
value: 66 66 66 66 66 66 66 66 66 66 66 66 66 66 66 66
       66 66 66 66 66 66 66 66 66 66 66 66 66 66 66 66
key: de ad be ef
value: f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0
       f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0
end

Wrote 64 bytes:
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
10 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 21 22 23 24 25 26 27 
28 29 2a 2b 2c 2d 2e 2f 
30 31 32 33 34 35 36 37 
38 39 3a 3b 3c 3d 3e 3f 

Read 64 bytes:
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
10 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 21 22 23 24 25 26 27 
28 29 2a 2b 2c 2d 2e 2f 
30 31 32 33 34 35 36 37 
38 39 3a 3b 3c 3d 3e 3f 

Wrote 1 bytes at pos 16 value ab.
Read 10 bytes with offset 10:
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
ab 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 21 22 23 24 25 26 27 
28 29 2a 2b 2c 2d 2e 2f 
30 31 32 33 34 35 36 37 
38 39 3a 3b 3c 3d 3e 3f 

Wrote 32 bytes at pos 32 with fancy enable mask.
Read 64 bytes with offset 0:
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
ab 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 ff 22 23 ff ff 26 27 
28 29 ff ff ff ff 2e 2f 
30 31 32 33 34 35 ff ff 
ff ff ff ff ff ff 3e 3f 

Read 0 bytes from map 66
Read 0 bytes from NULL key
Read 0 bytes to NULL data
Read 0 bytes from missing key
Read 0 bytes from too big key
Read 0 bytes from too small key
Read 64 bytes with request of 80 bytes
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
ab 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 ff 22 23 ff ff 26 27 
28 29 ff ff ff ff 2e 2f 
30 31 32 33 34 35 ff ff 
ff ff ff ff ff ff 3e 3f 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 

Read 32 bytes from offset 32 and requesting 64 bytes
20 ff 22 23 ff ff 26 27 
28 29 ff ff ff ff 2e 2f 
30 31 32 33 34 35 ff ff 
ff ff ff ff ff ff 3e 3f 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 

Wrote 0 bytes to map 66
Wrote 0 bytes to NULL key
Wrote 0 bytes of NULL data
Wrote 0 bytes with NULL enables
Wrote 0 bytes to large key
Wrote 0 bytes to small key
Wrote 0 bytes to offset 5 with req size 64 
Wrote 0 bytes with length 80
Updated 32 bytes at pos 0.
Read 64 bytes with offset 0:
f0 f0 f0 f0 f0 f0 f0 f0 
f0 f0 f0 f0 f0 f0 f0 f0 
f0 f0 f0 f0 f0 f0 f0 f0 
f0 f0 f0 f0 f0 f0 f0 f0 
20 ff 22 23 ff ff 26 27 
28 29 ff ff ff ff 2e 2f 
30 31 32 33 34 35 ff ff 
ff ff ff ff ff ff 3e 3f 

Updated 0 bytes at pos 0 with missing key.
Re-Updated 0 bytes at pos 0 with missing key.
Inserted 32 bytes at pos 0 with new key.
Read 64 bytes with offset 0:
f0 f0 f0 f0 f0 f0 f0 f0 
f0 f0 f0 f0 f0 f0 f0 f0 
f0 f0 f0 f0 f0 f0 f0 f0 
f0 f0 f0 f0 f0 f0 f0 f0 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 

Re-inserted 0 bytes at pos 0 with (not so) new key.
Removed newly inserted key, result 18446744073709551615.
Re-removed newly inserted key, result 0.
Read 0 bytes with offset 0 from removed key
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 

Final map state:
nanotube_map: 42 0 8 64
key: de ad be ef fe ed ba cc
value: f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0
       f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0 f0
       20 ff 22 23 ff ff 26 27 28 29 ff ff ff ff 2e 2f
       30 31 32 33 34 35 ff ff ff ff ff ff ff ff 3e 3f
end

Test passed.
